Switch over to ECMWF release 1.0.2 of ODC

srherbener opened the pull request:
This PR switches from the pre-release version of ODC to the official (version 1.0.2) release of ODC from ECMWF. The JCSDA repos have been renamed which makes this change pretty simple.

The repo we have had for a while with the pre-release version of ODC (version 0.99.0) has been renamed to JCSDA/odc-pre-release, and a new repo named JCSDA/odc has been built as a public fork of the public  ECMWF/odc repo.

In our fork, JCSDA/odc, a develop-jcsda branch has been created of which we will merge any necessary changes and also use to issue PRs with ECMWF. A tag "1.0.2.jcsda1" has been placed on the current develop-jcsda branch commit of which this PR now accesses in the build_stack.sh script.

mmiesch commented:
Thanks @srherbener for this update.  Is there a corresponding change also required for the odyssey build?

Also - in light of starting a new reporting period, we need to be more vigilant about checking all the PR metadata in code reviews.  I took the liberty of adding appropriate AOP20 epics

srherbener commented:
@mmiesch, there will be an update to odyssey. I thought I would do that separately since, I think, will be handled without a build process.

ECMWF discovered that the name odyssey was already taken on the PyPI repository, so they changed the name to pyodc. They made their repo public (ECMWF/pyodc) but also have posted the distribution package (from setup.py) on PyPI.

Since the distribution of pyodc is on PyPI, does that change how we handle it in jedi-stack? I was anticipating that we do a pip install in the environment, and then don't build pyodc during build_stack.sh, but perhaps I'm wrong about this.

mmiesch commented:
Thanks @srherbener - If it's just a pip install we can do it together with ```build_pyjedi.sh```.  I'll generate a new gnu-openmpi container for testing but before I rebuild all the containers, I'd like to wait until the pyodc bit is sorted out.

markjolah commented:
@srherbener I'm looking into this today.  I will be testing it with the HofX3D and ioda `feature/odc-writer` to read/write odc, and on multiple systems, so it may take a bit of time until I can confirm it is fully working.  I will also try adding `pyodc` to JEDI-Rapids dependencies which will install it via pip if not already present in the environment.

mmiesch commented:
There is now a docker image on Docker hub that has this new odc if anyone wants to try it:
~~~~~
docker pull jcsda/docker-gnu-openmpi-dev:odc
~~~~~
I'll make a singularity container later today.

mmiesch commented:
Singularity and charliecloud containers are now available in the usual places, with the name ``jedi-gnu-openmpi-dev-odc``.   Feel free to test if you wish.  Everything looks good to me but I haven't tested ioda-converters yet.

mmiesch commented:
Just a note on my last commit - previously, I was building odc and odyssey separately from the rest of the jedi-stack in the containers because it required github login credentials.  Now that it is public, it can be build with the rest.

srherbener commented:
@mmiesch, it turns out that the ``pip install`` doesn't work with pyodc. The package on PyPI is missing a file and one of the packages (codc) gets an error whey you try to import it.

I've added pyodc (using the JCSDA fork repo) to the build process in this PR and it is working on my Mac. I'll reset your review since this introduces a significant change.

Once things stabilize, I'll submit PRs to ECMWF with our local fixes.

mmiesch commented on modulefiles/mpi/compilerName/compilerVersion/mpiName/mpiVersion/pyodc/pyodc.lua:
```diff
@@ -0,0 +1,37 @@
+local opt = os.getenv("JEDI_OPT") or os.getenv("OPT") or "/opt/modules"
+local base = pathJoin(opt,compNameVerD,mpiNameVerD,pkgName,pkgVersion)
+local odc_lib = pathJoin(os.getenv("odc_ROOT"), "lib")
+local olib_pattern = pathJoin(base, "lib/python3*/site-packages")
```
Will this work right if someone has multiple python3* directories?

srherbener replied:
Good point. This would probably not work with multiple python3 directories. The odyssey lua script has the same issue, and this hasn't come up yet, but that might be because odyssey isn't used much and/or no one has upgraded their python3 installation yet.

I believe that what is desired is to select the lib/python3* directory that matches the version of python that is getting picked up on PATH. Is it possible to do this with lua?

mmiesch replied:
Actually, it does seem to work on Mac OSX.  I do have two matches to that capture query:
~~~~~
08:47 $ echo /usr/local/lib/python3*/site-packages
/usr/local/lib/python3.6/site-packages /usr/local/lib/python3.7/site-packages
~~~~~
I can run the build script successfully and when I load the module it looks like the capture command pulls out the first item in the list:
~~~~~
miesch:/opt/modules/modulefiles/core/jedi/gnu-openmpi$ echo $PYTHONPATH
/opt/modules/gnu-9.2.0/openmpi-4.0.1/pyodc/jcsda-1.0.0.jcsda1/lib/python3.6/site-packages:/opt/modules/gnu-9.2.0/openmpi-4.0.1/odc/jcsda-1.0.2.jcsda1/lib
~~~~~
And, it all seems to work even though my python executable is miniconda:
~~~~~
$ which python
/Users/miesch/miniconda3/bin/python
~~~~~
So, perhaps is it all ok and consistent.

markjolah commented:
@srherbener It may be easier to call pip install from a modified fork of pyodc with a working setup.py, instead of adding it to the modules system.  This will solve all of the problems, and we can push the necessary changes upstream as a PR.

markjolah commented:
Actually, your script already uses the setup.py from the actual ecmwf git repository which unlike the pypi archive actually does work.  You can use pip to directly install from a git repository instead of pypi.  Maybe that is all you need to do?

srherbener commented:
@markjolah, just so I'm clear, what you are suggesting is to remove the build_pyodc.sh script/process from jedi-stack and just have the user clone JCSDA/pyodc and run setup and pip (like what is being done in build_pyodc.sh). Ditto for the process that is building the python environment in the containers.

I'm okay doing it this way, @mmiesch does this sound okay with you? If so, where do I put the clone/setup.py/pip commands so the containers get pyodc installed?

Thanks!

mmiesch commented:
I think it is useful to still have the pyodc script, at least for the containers.  Otherwise I would have to duplicate the code there in all the different container recipe files and it would be inconvenient to maintain if something were to change.

mmiesch commented:
That said, we don't have to include it in the modules system.  We could just do a pip install in user space as is done in pyjedi.sh

srherbener commented:
@mmiesch, I'm a little surprised that it worked on your Mac when you have two lib/python3* paths. It seems that if your python is ``version 3.7.*``, then the build_stack.sh process would install pyodc into: ``/opt/modules/gnu-9.2.0/openmpi-4.0.1/pyodc/jcsda-1.0.0.jcsda1/lib/python3.7/site-packages``, and lua would place the ``python3.6`` version of that path on PYTHONPATH. If so, then python wouldn't be able to find pyodc, correct?

Anyway, it seems safer to move the clone/build/install of pyodc to the pyjedi.sh script (as is done for py-ncepbufr), and either build pyjedi or have the user do the clone/build/install manually. Would doing this cover the build of the containers, or should the build_pyodc.sh  and pyodc lua scripts be kept and shut off for non-container builds?

Thanks!

mmiesch commented:
Thanks @srherbener - So, my miniconda python is 3.6.9.  So, the pip install command installs it here ```/opt/modules/gnu-9.2.0/openmpi-4.0.1/pyodc/jcsda-1.0.0.jcsda1/lib/python3.6/site-packages```, which does get added to my path in the module build because it grabs the first item in the capture command.  But, this is probably just a fortuitous coincidence.

I think it would be best to do as you say - include it in pyjedi for now and do a standard pip install without specifying the prefix.  This should work for the containers and it should solve the problem noted above.

srherbener commented:
@mmiesch, @markjolah I have removed the build_pyodc.sh script and moved this functionality to the build_pyjedi.sh script. This way, the user can manually clone/build/install pyodc into their python installation, and build_pyjedi.sh can be used to automatically clone/build/install pyodc into the containers.

I tested this on my Mac, and it works.

Thanks for your feedback! I think this is a better configuration for pyodc than what I initially came up with.

mmiesch commented on buildscripts/libs/build_pyjedi.sh:
```diff
@@ -40,3 +40,15 @@ $SUDO python setup.py install
+cd pyodc
+git checkout 1.0.0.jcsda1
+
+python setup.py bdist_wheel
```
Should this be python3 as above?  It's probably good to be consistent either way

mmiesch commented:
Thanks @srherbener - I tested this on my Mac and in the gnu-openmpi-dev container.  Everything looks great.  If you want to test it, there is a docker container available called ```jcsda/docker-gnu-openmpi-dev:odc``` and a singularity container called ```library://jcsda/public/jedi-gnu-openmpi-dev:odc```

srherbener commented:
@mmiesch, I changed the python call in build_pyjedi.sh to "python3" as you suggested. I tried build_pyjedi.sh on my Mac and it worked. I also tried the odc Singularity container and that worked as well with the IDOA feature/odc-reader branch, and I was able to import pyodc and codc into python. Looks great, and thanks!

mmiesch commented:
Thanks @srherbener - I'll go ahead and merge this unless @markjolah wants to wait.  

I'd still like to merge [this PR](https://github.com/JCSDA/ecbuild/pull/28) before generating new containers, but then we're ready.

srherbener commented:
@mmiesch, sounds great. I'm good with waiting until the ecbuild PR is merged in before updating containers. Thanks!

srherbener commented:
@markjolah, are you okay with this PR now? It would be good to get this merged in soon. Thanks!

mmiesch commented:
@markjolah - ok if we merge this?

srherbener commented:
@mmiesch I've updated this PR to the latest ODC version from ecmwf (1.0.3). I tested on my MacBook and it successfully builds and I linked to the resulting ODC library in ufo-bundle and the IODA ODC tests pass.

The build in this PR uses the '1.0.3.jcsda1' tag in our fork of ecmwf's ODC repo. This tag is pointing to a fix in our fork that allows us to build with eckit 1.4. We will likely be okay using this until we can move to eckit 1.9, since we are not exercising a lot of features of ODC and we probably won't change things for a while.

I recommend merging now and moving forward with eckit 1.4, and syncing up with eckit 1.9 in a subsequent PR. If we run into a bug with ODC, the only recourse we have is to move to eckit 1.9 before ecmwf will look at it. That is true if we stay with what we have now (odb-api) or if we move to the released ODC libraries in this PR, so we might as well move to the released ODC libraries.

Does that sound good with you, and @markjolah?

Thanks!

mmiesch commented:
Thanks @srherbener - that sounds good to me.  As I mentioned yesterday,  there is a problem building eckit 1.10.1 with intel 19, but [they do have a fix](https://github.com/ecmwf/eckit/issues/27).  Their fix is to compile with `-fPIC`.  So, when we do upgrade to eckit >1.9 we should include this in the `build_eckit.sh` script.

mmiesch commented:
I don't mean to overburden this PR but while we're updating odc we may as well update eckit and ecbuild as well

srherbener commented:
@mmiesch, no problem with the eckit, ecbuild updates. I tested on my MacBook and was able to successfully build the stack, including odc version 1.0.3.

I had to change the following two specs in build_stack.sh:
* fckit
    * use jcsda feature/towards_sync_with_ecmwf branch
* atlas
    * use jcsda develop branch

I assume that fckit will get a tag once we are okay with the feature branch. Also, atlas got a build error using the ecmwf 0.19.1 version, which might be related to the eckit::Configuration update.

I also tested with ufo-bundle which built successfully, and all tests pass.

It looks like the install process for atlas is missing the Fortran modules. I had to build atlas as part of ufo-bundle to get access to the Fortran modules.

I'm good with tagging fckit feature/towards_sync_with_ecmwf now.

mmiesch commented on buildscripts/config/config_container-gnu-openmpi-dev.sh:
```diff
@@ -46,9 +46,8 @@ export    STACK_BUILD_ECBUILD=Y
-export        STACK_BUILD_ODB=N
 export        STACK_BUILD_ODC=Y
-export    STACK_BUILD_ODYSSEY=N
+export      STACK_BUILD_PYODC=N
```
We do want PYODC in the containers, don't we?

mmiesch replied:
Oh - sorry - I must have been looking at an old comment - disregard

srherbener replied:
Yes, but pyodc is part of the pyjedi package now since we eventually want to install directly from PyPI. This variable corresponds to a libs/build_pyodc.sh script, correct? If so, I need to remove these from the config files because they are not used anymore.

mmiesch commented:
> I had to change the following two specs in build_stack.sh:

I generally build fckit and atlas with the bundles 

> I assume that fckit will get a tag once we are okay with the feature branch. Also, atlas got a build error using the ecmwf 0.19.1 version, which might be related to the eckit::Configuration update.

> 
> I also tested with ufo-bundle which built successfully, and all tests pass.

Ame here - I tested fv3-bundle on my Mac and on AWS with gnu and intel - now working on S4.

> 
> It looks like the install process for atlas is missing the Fortran modules. I had to build atlas as part of ufo-bundle to get access to the Fortran modules.

Probably worth a zenhub issue although again, I don't think many people are installing atlas at the moment.
> 
> I'm good with tagging fckit feature/towards_sync_with_ecmwf now.

@benjaminmenetrier is making good progress here with Willem - I don't want to tag it while things are still changing.

srherbener commented:
@mmiesch, thanks for the info, and for testing and reviewing this PR.

So the normal operating mode is to exclude fckit and atlas from your stack build. Instead you include fckit and atlas in your bundle and build these as part of your bundle. If so, I'll go and update the config files to have fckit and atlas shut off by default.

Thanks!

mmiesch commented:
@srherbener @mer-a-o  - I just generated a new trio of containers (docker, singularity, charliecloud) that has all the latest updates included here, including ecbuild, eckit, odc, and pyodc.  It has a few other things as well, such as packages needed to generate atlas documentation (ford, imagemagick, tex4ht) and gnu time.  This is now the "latest" on docker hub so we'll see how it does with the CI in the next day or two.  If everything looks good I will proceed to generate the clang and intel containers.

srherbener commented:
I just committed a change that only affects the Mac configuration (config/config_mac.sh). I shut off the build of fckit and atlas for the Mac. These are turned off in all of the other configuration files. This change shouldn't affect the new containers that were recently built.

mmiesch commented:
Thanks @markjolah for the review - before we merge, I want to tag the previous version with release 0.1 so we can record what versions it uses of everything.
